op-e2e: Refactor interop fault proofs tests

ajsutton opened the pull request:
**Description**

Rough start of a DSL to make writing interop action tests easier and higher level.  This significantly reduces the amount of code required to setup the chain state needed for testing and generally makes the tests more readable. The code that actually runs op-program and challenger hasn't been pulled into the DSL yet - it should be in time, but was less of a maintenance headache than the setup code.

The DSL is currently specific to interop action tests and reuses the existing action test framework to minimise changes. I think it provides a good starting point for working out how we want to make tests easier to write but almost certainly isn't exactly what we'll land on. As we go forward we can refine the patterns used and should eventually move all e2e tests over to whatever the new style is that we settle on. I think we should merge this now because it's better than what's there and unblocks a lot of testing we've found very difficult for interop fault proofs.

As a follow up to this, I'll move the DSL into a separate package but that means moving the existing `interop.go` to a new package and that means changing a bunch of other tests so will avoid conflicts for now and leave it all in the one package.

Inphi commented on op-e2e/actions/interop/proofs_test.go:
```diff
@@ -0,0 +1,549 @@
+	dsl.AddL2Block(dsl.Actors.ChainA)
+	dsl.AddL2Block(dsl.Actors.ChainB)
+
+	dsl.AdvanceSafeHeads(func(opts *AdvanceSafeHeadOpts) {
+		opts.SetChains(dsl.Actors.ChainA)
+	})
+	dsl.AdvanceSafeHeads(func(opts *AdvanceSafeHeadOpts) {
+		opts.SetChains(dsl.Actors.ChainB)
+	})
```
nit: I think it can simplified as
```suggestion
	dsl.AdvanceSafeHeads(func(opts *AdvanceSafeHeadOpts) {
		opts.SetChains(dsl.Actors.ChainA, dsl.Actors.ChainB)
	})
```

ajsutton replied:
Ah, that was a deliberate change - we want the two chains to submit batches in separate L1 blocks (otherwise we could just use `dsl.AdvanceSafeHeads()` to submit batch data for all chains.  I probably should add a comment and the `AdvanceSafeHeads` method name might need a better name...

Inphi replied:
Yeah it wasn't clear to me it's submitting batches. The name suggests it's only updating derivation's view of L1 for batches that were already submitted.

ajsutton replied:
Went with `SubmitBatchData` and added a comment to explain why we are submitting the chains separately.

Inphi commented on op-e2e/actions/interop/proofs_test.go:
```diff
@@ -0,0 +1,549 @@
+	}
+
+	// Induce block replacement
+	//verifyCrossSafe(t, actors)
```
leftover comment

Inphi reviewed (approved):
Looks good. The tests are much more readable and hopefully they'll be easier to write.

ajsutton committed:
op-e2e: Refactor interop fault proofs tests

Rough start of a DSL to make writing interop action tests easier and higher level. This significantly reduces the amount of code required to setup the chain state needed for testing and generally makes the tests more readable.
